azuredevops/internal/service/permissions/utils/securityHelper.go

```diff
@@ -60,14 +65,33 @@ func GetPrincipalPermissions(d *schema.ResourceData, sn *SecurityNamespace, aclT
 	if err != nil {
 		return nil, err
 	}
-	if principalPermissions == nil || len(*principalPermissions) != 1 {
+	if principalPermissions == nil || len(*principalPermissions) <= 0 {
+		return nil, nil
+	}
+	if len(*principalPermissions) != 1 {
 		return nil, fmt.Errorf("Failed to retrieve current permissions for principal [%s]", principalList[0])
 	}
-	d.SetId(fmt.Sprintf("%s/%s", *aclToken, principal.(string)))
 	for key := range ((*principalPermissions)[0]).Permissions {
 		if _, ok := permissions.(map[string]interface{})[string(key)]; !ok {
 			delete(((*principalPermissions)[0]).Permissions, key)
 		}
 	}
 	return &(*principalPermissions)[0], nil
 }
+
+func InitializeSecurityNamespaceAndToken(d *schema.ResourceData, clients *client.AggregatedClient, namespaceID SecurityNamespaceID, tokenCreator func(d *schema.ResourceData, clients *client.AggregatedClient) (*string, error)) (*SecurityNamespace, *string, error) {
```

Comment:
I was thinking if we can define the `createToken` function as type somewhere like [service connection flatFunc ](https://github.com/terraform-providers/terraform-provider-azuredevops/blob/master/azuredevops/internal/service/serviceendpoint/commons.go#L19). Current  method params is too long.

Reply:
I took the chance to move the ACL token into the `SecurityNamespace` struct as an internal member, so the user of  the SecurityNamespace  doesn't have to handle it separately anymore. Made the interface and usage of the `SecurityNamespace` struct  even more clearer, saves some more code and reduces the number of parameters. 

In addition I added a function type for the ACL token creator.